### PR TITLE
Preserve all floating point registers

### PR DESCRIPTION
**Pull request by joshpeterson**
During a context switch, we should preserve all floating point
registers. With the new Mono compiler, we're seeing more sequence
points, which cause more context switches. If these registers are not
preserved, we can have incorrect float values returned when the debugger
is attached.

These changes correct Unity case 836825.

### CODE REVIEW

File: mono/mini/exceptions-amd64.c [hunk 1]
```diff
@@ -168,6 +168,38 @@ LONG CALLBACK seh_vectored_exception_handler(EXCEPTION_POINTERS* ep)
...
+	sctx->xmm11_low = ctx->Xmm11.Low;
+	sctx->xmm11_high = ctx->Xmm11.High;
+	sctx->xmm12_low = ctx->Xmm12.Low;
+	sctx->xmm12_high = ctx->Xmm12.High;
+	sctx->xmm13_low = ctx->Xmm13.Low;
+	sctx->xmm13_high = ctx->Xmm13.High;
+	sctx->xmm14_low = ctx->Xmm14.Low;
+	sctx->xmm14_high = ctx->Xmm14.High;
+	sctx->xmm15_low = ctx->Xmm15.Low;
+	sctx->xmm15_high = ctx->Xmm15.High;
```
Comment (joshpeterson) [inline mono/mini/exceptions-amd64.c]: There is lots of copy-paste here. I could instead write a few macros to hide this, and maybe make the code less error-prone. Is that worthwhile?

**Comment by joshpeterson** [comment]
@joncham You mentioned that upstream handles this better. Do we need to test and apply a similar fix upstream as well? Or is the issue not present there?

**Comment by joncham** [comment]
Yes, we should fix upstream as well. In upstream the context is just an array of values

https://github.com/mono/mono/blob/master/mono/utils/mono-context.h#L163

So we can just iterate and copy the fregs just like normal registers

https://github.com/mono/mono/blob/master/mono/mini/exceptions-amd64.c#L178

**Comment by joncham** [comment]
I think you want to have the other `#if` cases in `mono_arch_sigctx_to_monoctx` and `mono_arch_monoctx_to_sigctx` for non-Windows platforms. At least whatever path we take on OSX.

**Comment by joshpeterson** [comment]
I was able to reproduce case 836825 on OSX and confirm that the latest change corrects the issue. I'm running a full Mono build to make sure that I haven't broken anything on any other platforms now.

**Comment by joshpeterson** [comment]
The Mono build at the latest commit on this branch is green. Let me know if you think I can merge this.

**Comment by joncham** [comment]
LGTM

**Pull request merged**